teapots/choreographer-30fps/src/main/cpp/CMakeLists.txt

```diff
@@ -17,36 +17,53 @@
 cmake_minimum_required(VERSION 3.4.1)
 
 # build native_app_glue as a static lib
-add_library(native_app_glue STATIC
-    ${ANDROID_NDK}/sources/android/native_app_glue/android_native_app_glue.c)
-
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++11 -Wall -fno-exceptions -fno-rtti")
+add_library(NativeAppGlue STATIC
+  ${ANDROID_NDK}/sources/android/native_app_glue/android_native_app_glue.c)
 
 # build the ndk-helper library
-set(ndk_helper_dir ../../../../common/ndk_helper)
-add_subdirectory(${ndk_helper_dir} ndk_helper)
-
-# Export ANativeActivity_onCreate(), 
-# Refer to: https://github.com/android-ndk/ndk/issues/381.
-set(CMAKE_SHARED_LINKER_FLAGS
-    "${CMAKE_SHARED_LINKER_FLAGS} -u ANativeActivity_onCreate")
+set(ndkHelperSrc ../../../../common/ndk_helper)
+add_subdirectory(${ndkHelperSrc} ndkHelperBin)
```

Comment:
A cool trick that helps with large projects:

add_subdirectory(${ndkHelperSrc} ndkHelperBin) can be changed to
add_subdirectory(${ndkHelperSrc} ${CMAKE_CURRENT_BINARY_DIR}/ndkHelperBin)
The difference is subtle but important.
AS will do a different build per variant + per flavor + per build type.  So a good acceleration is to specify that the gradle should build to a folder other than .externalNativeBuild and reuse build bin paths.  This means the trick is to tell cmake that it should create a custom bin path and use that everywhere.  To take advantage of it, our binary obj build path needs to exist under the current bin directory.

Reply:
that is good one as building time is import! one more step:  as this ndkHelper module is supporting 3 projects, each CMake_CURRENT_BINARY_DIR will not be the same, so it will still build 3 times, correct?  What happen if I hardcode to a common directory as bin, then 3 project will be able to use it, instead of building it 3 times, correct?

Reply:
Yep,  If you hardcode it to be a certain path then CMake is smart enough to decide that the output already exists and thus wont need rebuilding.